### PR TITLE
Prefer peer with good response rate

### PR DESCRIPTION
**Pull request by zilm13**
## PR Description
Just a very simple solution. It will loose status on peer disconnection. But it's simple and should work. Works well in my tests.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

---

> [!NOTE]
> **Medium Risk**
> Changes the peer-selection heuristic for data column sidecar retrieval and introduces new per-peer success metrics, which could affect request distribution and retrieval liveness under real network conditions.
> 
> **Overview**
> Updates `SimpleSidecarRetriever` peer selection to **prefer peers with the highest historical sidecar success rate** (`ConnectedPeer.getResponseScore()`), using available request capacity only as a tiebreaker.
> 
> Adds per-peer requested/received counters (with overflow reset) and hooks them into request start/success paths, removes per-request peer request-count tracking, and exposes `getConnectedPeers()` for testing.
> 
> Refreshes tests and fixtures to reflect the new behavior: `TestPeer`/`TestPeerManager` now track *available* request capacity via completed requests, and new/updated tests verify request-limit handling and response-score-driven peer preference.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ad128e96098bb1cb5ee312f043b4905ef7c5323b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 1]
```diff
@@ -287,26 +290,20 @@ private record ActiveRequest(SafeFuture<Void> promise, ConnectedPeer peer) {}
...
-      return peerRequestCount.getOrDefault(peerId, 0);
-    }
   }
 
   private class ConnectedPeer {
     final UInt256 nodeId;
     final Cache<CacheKey, Set<UInt64>> custodyIndicesCache = LRUCache.create(2);
+    // just to avoid starting with non-divisible 0/0
+    final AtomicInteger sidecarsRequested = new AtomicInteger(1);
+    final AtomicInteger sidecarsReceived = new AtomicInteger(1);
```
Comment (rolfyone) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java]: these numbers could get big, maybe we just use AtomicLong?  some people leave their boxes running for months and months so we probably need overflow protection where we re-initialise both if they happen to overflow? if we did overflow protection could stay as AtomicInteger...

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 2]
```diff
@@ -144,14 +144,15 @@ private boolean activateMatchedRequest(final RequestMatch match) {
 
     final SafeFuture<DataColumnSidecar> reqRespPromise =
         reqResp.requestDataColumnSidecar(match.peer.nodeId, match.request.columnId);
-    match.request.onPeerRequest(match.peer().nodeId);
+    match.peer.sidecarsRequested.getAndIncrement();
```
Comment (rolfyone) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java]: basically if we re-initialise we can just warp increments and catch overflow and call an init fn that sets both to 1... also this should probably be incrementAndGet, as a nit...

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 3]
```diff
@@ -144,14 +144,15 @@ private boolean activateMatchedRequest(final RequestMatch match) {
...
 
     final SafeFuture<Void> activeRpcRequest =
         reqRespPromise.handle(
             (sidecar, err) -> {
               reqRespCompleted(match.request, sidecar);
-
-              if (err != null) {
+              if (err == null) {
+                match.peer.sidecarsReceived.incrementAndGet();
```
Comment (rolfyone) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java]: same theory, catch and re-initialise here.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 4]
```diff
@@ -175,7 +176,9 @@ private Optional<ConnectedPeer> findBestMatchingPeer(
 
     // taking first the peers which were not requested yet, then peers which are less busy
     final Comparator<ConnectedPeer> comparator =
-        Comparator.comparing((ConnectedPeer peer) -> request.getPeerRequestCount(peer.nodeId))
+        Comparator.comparing(
+                (ConnectedPeer peer) ->
+                    (float) peer.sidecarsReceived.get() / peer.sidecarsRequested.get())
```
Comment (rolfyone) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java]: id be tempted to have this in the class rather than accessing private attributes here, would also let you test edge cases if we ended up testing that internal class at some point...

could have something like a `sidecarRequestSuccessRate` in `ConnectedPeer`

**Comment by zilm13** [comment]
@rolfyone 
Addressed feedback:
  - kept `AtomicInteger` but added frontend to handle overflow
  - added frontend for success rate score to avoid calculating it outside ConnectedPeer

Also added tests to clarify logic is good (it was not actually, so fixed 1 bug, now it's good)

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 5]
```diff
@@ -321,13 +340,47 @@ private Set<UInt64> calcNodeCustodyIndices(final CacheKey cacheKey) {
...
+      }
+    }
+
+    public void countSidecarReceived() {
+      final int current = sidecarsReceived.incrementAndGet();
+      if (current == Integer.MAX_VALUE) {
+        sidecarsRequested.set(1);
+        sidecarsReceived.set(1);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (zilm13) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java:388]: Theoretically yes, but rarely and not the end of the world, i think. The fix will require synchronization for this method which destroys whole atomic counters idea.

[the remark this answers is not in the document]
Comment (rolfyone) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java:388]: im still hopeful we dont even need that reset code, but in the case we do, we're better off having it than not, and missing potentially 1 value im not concerned about.

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetrieverTest.java [hunk 6]
```diff
@@ -299,4 +407,29 @@ void shouldTrackCustodyCountChangesForPeers() {
...
+    assertThat(connectedPeer.getResponseScore()).isEqualTo(1);
+    connectedPeer.getSidecarsRequested().incrementAndGet();
+    assertThat(connectedPeer.getResponseScore()).isEqualTo(0);
+    connectedPeer.getSidecarsReceived().addAndGet(3);
+    assertThat(connectedPeer.getResponseScore()).isEqualTo(3);
+    connectedPeer.getSidecarsRequested().addAndGet(1_000_000);
+    assertThat(connectedPeer.getResponseScore()).isEqualTo(0);
+    connectedPeer.getSidecarsReceived().addAndGet(1_000_000);
+    assertThat(connectedPeer.getResponseScore()).isEqualTo(9);
+  }
```
Comment (rolfyone) [inline ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetrieverTest.java]: these kids of tests are much much cleaner if they're not doing a bunch of things.... one simple option is a parameterised test where you have each `requested`, `received`, `score` parameters and then it becomes readable...

Comment (zilm13) [reply ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetrieverTest.java]: Makes sense! Updated test to verify atomic changes

File: ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetrieverTest.java [hunk 7]
```diff
@@ -210,6 +232,95 @@ void selectingBestPeerShouldRespectPeerMetrics() {
     assertThat(allRequestCountsFunc.get()).isEqualTo(List.of(0, 1, 2, 2));
   }
 
+  @Test
+  void selectingBestPeerShouldRespectResponseScore() {
```
Comment (rolfyone) [inline ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetrieverTest.java]: tests like this are very hard to follow, a unit test generally is a lot clearer to read and reason on.

id suggest breaking this down into more atomic things that are showing what you want to show, rather than testing all things in one long fn.

eg theres no reason why we couldn't inject a starting state of a connectedPeer, and then based on that show pretty much any scenario in a few lines.

Comment (zilm13) [reply ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetrieverTest.java]: I have refactored it, so I've added several helpers and I think now tests are way more readable and you could focus more on scenario. Also I've dropped second part of this test, because it's a duplicated check, we ensure that request limit is used in peer choice by the test `selectingBestPeerShouldRespectRequestLimits`

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 8]
```diff
@@ -125,7 +126,8 @@ public void start() {}
   @Override
   public void stop() {}
 
-  private Stream<RequestMatch> matchRequestsAndPeers() {
+  @VisibleForTesting
+  Stream<RequestMatch> matchRequestsAndPeers() {
```
[the remark this answers is not in the document]
Comment (zilm13) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java]: Nice catch, done

File: ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/retriever/TestPeer.java [hunk 9]
```diff
@@ -114,6 +120,14 @@ public int getCurrentRequestLimit() {
     return currentRequestLimit;
   }
 
+  public int getAvailableRequestCount() {
```
[the remark this answers is not in the document]
Comment (zilm13) [reply ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/retriever/TestPeer.java:123]: I'd keep it as it's a base getter.

**Review by rolfyone — APPROVED** [review]
LGTM

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 10]
```diff
@@ -342,6 +358,41 @@ public boolean hasSlotAvailable(final UInt64 slot) {
...
+     */
+    public int getResponseScore() {
+      return (int) (sidecarsReceived.get() * 10.0F / sidecarsRequested.get());
+    }
+
+    public void countSidecarRequest() {
+      final int current = sidecarsRequested.incrementAndGet();
+      if (current == Integer.MAX_VALUE) {
+        sidecarsRequested.set(1);
+        sidecarsReceived.set(1);
```
Comment (tbenr) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java]: a nit: we could have a `resetCounters` and put these two set in there

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java [hunk 11]
```diff
@@ -178,8 +180,7 @@ private Optional<ConnectedPeer> findBestMatchingPeer(
 
     // taking first the peers which were not requested yet, then peers which are less busy
```
Comment (tbenr) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java]: this is comment is not valid anymore.

Comment (tbenr) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java]: AI says something like: `preferring peers with the best response rate, then preferring less busy peers among equals.`

**Review by tbenr — APPROVED** [review]
LGTM, I wasn't able to test in the conditions I wanted, but I tested locally and looks good
